src/linear_algebra/affine_space.lean

```diff
@@ -0,0 +1,235 @@
+/-- All points are in `affine_subspace.univ`. -/
+lemma affine_subspace.mem_univ (p : P) : p ∈ affine_subspace.univ k V P :=
+set.mem_univ p
+
+instance : inhabited (affine_subspace k V P) := ⟨affine_subspace.univ k V P⟩
+
+/-- The affine span of a nonempty set of points is the smallest affine
+subspace containing those points. (Actually defined here in terms of
+spans in modules.) -/
+def affine_span (s : set P) (h : s.nonempty) : affine_subspace k V P :=
+{ carrier := span_points k V s,
+  direction := vector_span k V s,
+  nonempty := span_points_nonempty_of_nonempty k V h,
+  add := λ p v hp hv, vadd_mem_span_points_of_mem_span_points_of_mem_vector_span k V hp hv,
+  sub := λ p1 p2 hp1 hp2, vsub_mem_vector_span_of_mem_span_points_of_mem_span_points k V hp1 hp2 }
+
+/-- The affine span, converted to a set, is `span_points`. -/
+@[simp] lemma affine_span_coe (s : set P) (h : s.nonempty) :
+  (affine_span k V P s h : set P) = span_points k V s :=
+rfl
+
+/-- A point in a set is in its affine span. -/
+lemma affine_span_mem (p : P) (s : set P) (hp : p ∈ s) : p ∈ affine_span k V P s ⟨p, hp⟩ :=
+mem_span_points k V p s hp
+
+end affine_subspace
+
+section affine_map
+
+variables  (k : Type*) (V1 : Type*) (P1 : Type*) (V2 : Type*) (P2 : Type*)
+    (V3 : Type*) (P3 : Type*) [ring k]
+    [add_comm_group V1] [module k V1] [affine_space k V1 P1]
+    [add_comm_group V2] [module k V2] [affine_space k V2 P2]
+    [add_comm_group V3] [module k V3] [affine_space k V3 P3]
+
+/-- An `affine_map k V1 P1 V2 P2` is a map from `P1` to `P2` that
+induces a corresponding linear map from `V1` to `V2`. -/
+structure affine_map :=
+(to_fun : P1 → P2)
+(linear : linear_map k V1 V2)
+(add : ∀ (p : P1) (v : V1), to_fun (v +ᵥ p) =  linear.to_fun v +ᵥ to_fun p)
+
+instance: has_coe_to_fun (affine_map k V1 P1 V2 P2) := ⟨_, affine_map.to_fun⟩
+
+/-- Constructing an affine map and coercing back to a function
+produces the same map. -/
+@[simp] lemma affine_map.coe_mk (f : P1 → P2) (linear add) :
+  ((affine_map.mk f linear add : affine_map k V1 P1 V2 P2) : P1 → P2) = f := rfl
+
+/-- `to_fun` is the same as the result of coercing to a function. -/
+@[simp] lemma affine_map.to_fun_eq_coe (f : affine_map k V1 P1 V2 P2) : f.to_fun = ⇑f := rfl
+
+/-- An affine map on the result of adding a vector to a point produces
+the same result as the linear map applied to that vector, added to the
+affine map applied to that point. -/
+@[simp] lemma affine_map.map_vadd (f : affine_map k V1 P1 V2 P2) (p : P1) (v : V1) :
+  f (v +ᵥ p) = f.linear v +ᵥ f p := f.add p v
+
+/-- The linear map on the result of subtracting two points is the
+result of subtracting the result of the affine map on those two
+points. -/
+@[simp] lemma affine_map.map_vsub (f : affine_map k V1 P1 V2 P2) (p1 p2 : P1) :
+  f p1 -ᵥ f p2 = f.linear (p1 -ᵥ p2) :=
+by conv_lhs { rw [←vsub_vadd V1 p1 p2, affine_map.map_vadd, vadd_vsub] }
+
+/-- Identity map as an affine map. -/
+def affine_map.id : affine_map k V1 P1 V1 P1 :=
```

Comment:
You need a `@[simp]` lemma saying what `affine_map.id` does when applied to a point.

Reply:
Added `affine_map.id_apply`.